core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,552 @@
+        driver.reconcileTasks(statuses)
+      }
+    }
+  }
+
+  private def buildCommand(req: DriverSubmission): CommandInfo = {
+    val desc = req.desc
+
+    val appJar = CommandInfo.URI.newBuilder()
+      .setValue(desc.desc.jarUrl.stripPrefix("file:").stripPrefix("local:")).build()
+
+    val builder = CommandInfo.newBuilder()
+      .addUris(appJar)
+
+    val entries =
+      (conf.getOption("spark.executor.extraLibraryPath").toList ++
+        desc.desc.command.libraryPathEntries)
+
+    val prefixEnv = if (!entries.isEmpty) {
+      Utils.libraryPathEnvPrefix(entries)
+    } else {
+      ""
+    }
+
+    val envBuilder = Environment.newBuilder()
+    desc.desc.command.environment.foreach {
+      case (k, v) =>
+        envBuilder.addVariables(
+          Variable.newBuilder().setName(k).setValue(v).build())
+    }
+
+    builder.setEnvironment(envBuilder.build())
+
+    val cmdOptions = generateCmdOption(req)
+
+    val executorUri = req.desc.schedulerProperties.get("spark.executor.uri")
+      .orElse(req.desc.desc.command.environment.get("SPARK_EXECUTOR_URI"))
+
+    val cmd = if (executorUri.isDefined) {
+      builder.addUris(CommandInfo.URI.newBuilder().setValue(executorUri.get).build())
+
+      val folderBasename = executorUri.get.split('/').last.split('.').head
+
+      val cmdExecutable = s"cd $folderBasename*; $prefixEnv bin/spark-submit"
+
+      val cmdJar = s"../${desc.desc.jarUrl.split("/").last}"
+
+      val appArguments = desc.desc.command.arguments.mkString(" ")
+
+      s"$cmdExecutable ${cmdOptions.mkString(" ")} $cmdJar $appArguments"
+    } else {
+      val executorSparkHome = req.desc.schedulerProperties.get("spark.mesos.executor.home")
+        .orElse(conf.getOption("spark.home"))
+        .orElse(Option(System.getenv("SPARK_HOME")))
+        .getOrElse {
+          throw new SparkException("Executor Spark home `spark.mesos.executor.home` is not set!")
+        }
+
+      val cmdExecutable = new File(executorSparkHome, "./bin/spark-submit").getCanonicalPath
+
+      val cmdJar = desc.desc.jarUrl.split("/").last
+
+      s"$cmdExecutable ${cmdOptions.mkString(" ")} $cmdJar"
+    }
+
+    builder.setValue(cmd)
+
+    builder.build
+  }
+
+  private def generateCmdOption(req: DriverSubmission): Seq[String] = {
+    var options = Seq(
+        "--name", req.desc.schedulerProperties("spark.app.name"),
+        "--class", req.desc.desc.command.mainClass,
+        "--master", s"mesos://${conf.get("spark.master")}",
+        "--driver-cores", req.desc.desc.cores.toString,
+        "--driver-memory", s"${req.desc.desc.mem}M")
+
+    req.desc.schedulerProperties.get("spark.executor.memory").map { v =>
+      options ++= Seq("--executor-memory", v)
+    }
+
+    req.desc.schedulerProperties.get("spark.cores.max").map { v =>
+      options ++= Seq("--total-executor-cores", v)
+    }
+
+    options
+  }
+
+  private [spark] case class ResourceOffer(val offer: Offer, var cpu: Double, var mem: Double)
```

Comment:
why is this `private[spark]`? This is only used in this file. Also, case classes by definition should be immutable. If it takes in a var then it should prob not be a case class.